src/matchers.ts

```diff
@@ -56,3 +62,146 @@ export function isAllOf<Matchers extends [Matcher<any>, ...Matcher<any>[]]>(
+  return hasExpectedRequestMetadata(action, ['pending'])
+}
+
+export type UnknownAsyncThunkRejectedAction = ReturnType<
+  AsyncThunkRejectedActionCreator<unknown, unknown>
+>
+
+/**
+ * A type guard function that matches any rejected async thunk action.
+ *
+ * @param action A redux action
+ *
+ * @public
+ */
+export function isRejected(
+  action: any
+): action is UnknownAsyncThunkRejectedAction {
+  return hasExpectedRequestMetadata(action, ['rejected'])
+}
+
+export type UnknownAsyncThunkFulfilledAction = ReturnType<
+  AsyncThunkFulfilledActionCreator<unknown, unknown>
+>
+
+/**
+ * A type guard function that matches any fulfilled async thunk action.
+ *
+ * @param action A redux action
+ *
+ * @public
+ */
+export function isFulfilled(
+  action: any
+): action is UnknownAsyncThunkFulfilledAction {
+  return hasExpectedRequestMetadata(action, ['fulfilled'])
+}
+
+export type UnknownAsyncThunkAction =
+  | UnknownAsyncThunkPendingAction
+  | UnknownAsyncThunkRejectedAction
+  | UnknownAsyncThunkFulfilledAction
+
+/**
+ * A type guard function that matches any async thunk action.
+ *
+ * @param action
+ *
+ * @internal
+ */
+export function isAnyAsyncThunkAction(
+  action: any
+): action is UnknownAsyncThunkAction {
+  return hasExpectedRequestMetadata(action, [
+    'pending',
+    'fulfilled',
+    'rejected'
+  ])
+}
+
+export type AnyAsyncThunk = AsyncThunk<any, any, any>
+
+export type ActionsFromAsyncThunk<T extends AnyAsyncThunk> =
+  | ActionFromMatcher<T['pending']>
+  | ActionFromMatcher<T['fulfilled']>
+  | ActionFromMatcher<T['rejected']>
+
+/**
+ * A higher-order function that returns a function that may be used to check
+ * whether an action belongs to one of the provided async thunk action creators.
+ *
+ * @param asyncThunks (optional) The async thunk action creators to match against.
+ *
+ * @public
+ */
+export function isAsyncThunkAction(): (
+  action: any
+) => action is UnknownAsyncThunkAction
+export function isAsyncThunkAction<
+  AsyncThunks extends [AnyAsyncThunk, ...AnyAsyncThunk[]]
+>(
+  ...asyncThunks: AsyncThunks
+): (action: any) => action is ActionsFromAsyncThunk<AsyncThunks[number]>
+export function isAsyncThunkAction<
+  AsyncThunks extends [AnyAsyncThunk, ...AnyAsyncThunk[]]
+>(...asyncThunks: AsyncThunks) {
+  if (asyncThunks.length === 0) {
+    return isAnyAsyncThunkAction
+  }
+
+  return (
+    action: any
+  ): action is ActionsFromAsyncThunk<AsyncThunks[number]> => {
+    const [firstAsyncThunk, ...restAsyncThunks] = asyncThunks
+
+    const matchers: [Matcher<any>, ...Matcher<any>[]] = [
+      firstAsyncThunk.pending,
+      firstAsyncThunk.rejected,
+      firstAsyncThunk.fulfilled
+    ]
+
+    for (const asyncThunk of restAsyncThunks) {
+      matchers.push(
+        asyncThunk.pending,
+        asyncThunk.rejected,
+        asyncThunk.fulfilled
+      )
+    }
+
+    const combinedMatcher = isAllOf(isAnyAsyncThunkAction, isAnyOf(...matchers))
+    return combinedMatcher(action)
```

Comment:
This seems a bit long. I *think* this should do it as well, right?

```suggestion
    const matchers: [Matcher<any>, ...Matcher<any>[]] = []

    for (const asyncThunk of asyncThunks) {
      matchers.push(
        asyncThunk.pending,
        asyncThunk.rejected,
        asyncThunk.fulfilled
      )
    }

    const combinedMatcher = isAnyOf(...matchers)
    return combinedMatcher(action)
```

Reply:
@phryneas The problem was that `const matchers: [Matcher<any>, ...Matcher<any>[]] = []` was failing type checking because [] does not have at least one element.  Specifically, `"Property '0' is missing in type [] but required in ...`.

And if I change the annotation to `const matchers: Matcher<any>[]`, then `isAnyOf(...matchers)` fails type checking because at least one argument is required.

I'm also not completely satisfied with how the code looks.  It's organized this way to pass type checking without using any casts.  Maybe there's a more straightforward way?

Reply:
Honestly, I'm totally fine with a
`const matchers: [Matcher<any>, ...Matcher<any>[]] = [] as unknown`
there. After all, we know better than the compiler.